pulp_python/app/pypi.py

```diff
@@ -0,0 +1,195 @@
+from rest_framework.reverse import reverse
+from django.contrib.sessions.models import Session
+from django.db import transaction
+from django.db.models import Count
+from django.db.utils import DatabaseError, IntegrityError
+from django.shortcuts import redirect, get_object_or_404
+from django.http.response import Http404, HttpResponseForbidden, HttpResponseBadRequest, HttpResponse, StreamingHttpResponse
+from drf_spectacular.utils import extend_schema
+from dynaconf import settings
+from urllib.parse import urljoin
+from pathlib import PurePath
+from .utils import python_content_to_json, PYPI_LAST_SERIAL, PYPI_SERIAL_CONSTANT
+
+from pulpcore.plugin.tasking import dispatch
+from pulpcore.app.models import Artifact
+from pulp_python.app.models import (
+    PythonDistribution,
+    PythonPackageContent,
+    PythonPublication,
+    PythonRepository,
+)
+from pulp_python.app.serializers import PythonPackageContentSerializer, MinimalPythonPackageContentSerializer, PyPIPackageUploadSerializer
+from pulp_python.app.utils import write_simple_index, write_simple_detail
+from pulp_python.app import tasks
+
+log = logging.getLogger(__name__)
+
+BASE_CONTENT_URL = urljoin(settings.CONTENT_ORIGIN, settings.CONTENT_PATH_PREFIX)
+
+class PyPIMixin:
+    """Mixin to get index specific info."""
+    @staticmethod
+    def get_repository_version(distribution):
+        pub = distribution.publication
+        rep = distribution.repository
+        if pub:
+            return pub.repository_version or pub.repository.latest_version()
+        elif rep:
+            return rep.latest_version()
+        else:
+            raise Http404("No repository associated with this index")
+
+    @staticmethod
+    def get_distribution(path):
+        return get_object_or_404(PythonDistribution, base_path=path)
+
+    @staticmethod
+    def get_content(repository_version):
+        return PythonPackageContent.objects.filter(pk__in=repository_version.content)
+
+    def get_drvc(self, path):
+        distro = self.get_distribution(path)
+        repo_ver = self.get_repository_version(distro)
+        content = self.get_content(repo_ver)
+        return distro, repo_ver, content
+
+
+class PackageUploadMixin(PyPIMixin):
+
+    @extend_schema(request=PyPIPackageUploadSerializer)
+    def upload(self, request, path):
+        """
+        Upload a package to the index.
+
+        0. Check response is in correct format
+        1. Check if the index allows uploaded packages (live-api-enabled)
+        2. Check if the package is in the repository already
+        3. If present then reject request
+        4. Spawn task to add content if no/old session present
+        5. Add uploads to current session to group into one task
+        """
+        serializer = PyPIPackageUploadSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        file, sha256 = serializer.validated_data["content"], serializer.validated_data["sha256_digest"]
```

Comment:
FTR warehouse stores multiple hashes if you want to replicate that https://pypi.org/project/build/#copy-hash-modal-9cd0d632-b45a-435a-8d43-1b36ecdbd951

Reply:
For every artifact in Pulp we store multiple hashes https://github.com/pulp/pulpcore/blob/master/pulpcore/app/models/content.py#L142, but sha256 is the main one we use.  We disabled by default using md5 for security purposes and we don't calculate or store blake256 digests.  I don't think it's necessary we check all 3 since sha256 is pretty secure and all the tools upload the package with the sha256.